### DISCUSSION TITLE
[Question] Is it possible/advisable to run Fooocus in production as an API endpoint?

### DISCUSSION THREAD

**Discussion by TowhidKashem** [lllyasviel/Fooocus#2772] [Q&A] [answered]
Hi,

First thank you so much for making this! I was about to give up on image generation until I ran into Fooocus, it's so easy to do stuff like consistent characters that was going way over my head with automatic and comfyui! Also really appreciate the awesome defaults so I don't need to be a wizard to make nice pics!

My question is about exposing an endpoint for users in my app to generate their own images. Does fooocus run in an API mode? or have an SDK which I can host using a python or node server or the like? I guess I just need to be able to run stable diffusion through fooocus programmatically from code to make this endpoint. Was wondering if there are any guides or if it's even recommended for production use and not just personal use?

**Comment by mashb1t** [comment, marked as answer]
@TowhidKashem Fooocus can be used as API, see https://github.com/lllyasviel/Fooocus/discussions?discussions_q=is%3Aopen+api (example in https://github.com/lllyasviel/Fooocus/discussions/2608 or https://github.com/toutjavascript/FoooXus-Fooocus-Extender/blob/main/src/api.py#L222).

It comes with a queueing system by default and can handle users independently since [v2.2.0](https://github.com/lllyasviel/Fooocus/releases/tag/2.2.0), more specifically https://github.com/lllyasviel/Fooocus/pull/981.
Please be advised that you might encounter issues as the API uses Gradio UI controls as parameters and basically does the same as when you click a button in the FE (UI changes on updates => API changes), and is Websockets based, not REST.

Depending on what "production" means you'll also have to make sure you're compliant with [GPLv3](https://www.gnu.org/licenses/gpl-3.0.en.html) licensing. You might also switch from `entrypoint_with_update.py` to `launch.py` to not accidentally get unexpected updates in the future.

For maximum performance, please check out the options `--always-high-vram --disable-offload-from-vram` (also see Colab testing in https://github.com/lllyasviel/Fooocus/pull/1710#issuecomment-1989185346). It is in general recommended to load all models in GPU when sufficient VRAM is available (> 14GB), but please test yourself. Fooocus might still crash when insufficient resources are available on the hardware it's running on.

If you've considered all mentioned things you're ready to go. Hope this helps and happy implementation!

**Reply by TowhidKashem** [reply]
Awesome thank you, I'll read through all that!

One more thing on the topic if you don't mind...and I may be way over simplifying things here but my initial thought is do I even need Fooocus as an api layer, since all it does is tweak the options that are passed to stable diffusion, if there was some way to get the final "output command" that Fooocus produces before running SD, then shouldn't it be possible to bypass it entirely?

In my app the user just needs to enter a prompt, the settings aren't changeable by them. I want to use the Fooocus UI to determine the best settings for the type of images I want them to create and hard code these on the backend. So for me is the Fooocus layer necessary? I just need to be able to get the final SD options that Foocus produces just once right? Then I can use any stable diffusion endpoint and pass the exact same options and get the same type of images each time right? If so, is there a way to get these raw SD options created by Fooocus somehow ? 

I saw that the "History log" shows all the images and underneath you can copy the Foocus payloads for example:

```
{
"prompt": "a woman, New York city",
"negative_prompt": "unrealistic, saturated, high contrast, big nose, painting, drawing, sketch, cartoon, anime, manga, render, CG, 3d, watermark, signature, label",
"prompt_expansion": "a woman, New York city, sharp focus, intricate, elegant, highly detailed, cinematic, candid, beautiful, bright, romantic, thought, dramatic ambient, sublime, symmetry, epic, iconic, amazing, gorgeous, colorful, pretty, fine, complex, elaborate, color, awesome, creative, positive, pure, coherent, attractive, loving, peaceful, cute, enhanced",
"styles": "['Fooocus V2', 'Fooocus Photograph', 'Fooocus Negative']",
"performance": "Speed",
"resolution": "(896, 1152)",
"guidance_scale": 3,
"sharpness": 2,
"adm_guidance": "(1.5, 0.8, 0.3)",
"base_model": "realisticStockPhoto_v20.safetensors",
"refiner_model": "None",
"refiner_switch": 0.5,
"sampler": "dpmpp_2m_sde_gpu",
"scheduler": "karras",
"seed": "4259690989367531913",
"lora_combined_1": "SDXL_FILM_PHOTOGRAPHY_STYLE_BetaV0.4.safetensors : 0.25",
"metadata_scheme": false,
"version": "Fooocus v2.3.1"
}
```

But these are Foocus specific from the looks of it and can't b used in SD as is right?

**Reply by mashb1t** [reply]
In general you're right, you could just use [diffusers](https://huggingface.co/docs/diffusers/using-diffusers/sdxl) as wrapper, but you need a wrapper anyways, might as well be Fooocus.

Fooocus does much more than just wrap around SD, it uses separate pre- and post-processing as well as memory and other performance and quality optimisations. There are no raw options you could extract, as also core functionality is overridden and optimized.

Feel free to check the diffusers docs and compare the quality of both.

EDIT compared to A1111 (also Forge) and ComfyUI, Fooocus is the most static but also offers the best and tightly integrated OOTB pipeline to achieve superior performance and quality (depending on workflows and extensions of the other tools).

**Reply by TowhidKashem** [reply]
gotcha thank you! I wasn't aware of any of this so didn't know I needed the wrapper. It sounds like fooocus is production ready and those out of the box optimizations sounds nice. Will give it a shot after reading some more, thanks again!

**Reply by rishabh063** [reply]
@mashb1t I was looking into this i need to optimise my production pipeline and do just a few things not all , is there a way i can load a striped up version of fooocus ( like load fewer models etc )

**Reply by mashb1t** [reply]
@rishabh063 Fooocus will only install the models which are needed to fulfill a specific request and does not download all by default => is by design as lean as possible.

**Reply by rishabh063** [reply]
For example i only need to outpaint can i strip out gpt2 model ?

**Reply by mashb1t** [reply]
yes, you can simply not use the style "Fooocus V2"
